Automattic/newspack-plugin
# Wizard and WizardScreen components for managing common wizard tasks

### All Submissions:






### Changes proposed in this Pull Request:



This PR adds `Wizard` and `WizardScreen` components and migrates the `Subscriptions` and `SubscriptionsOnboarding` wizards to use those components. These components should let us build wizards without needing to re-implement a bunch of common functionality each time.

- The `WizardScreen` component renders one screen of a wizard, with optional primary button and optional tertiary button.
- The `Wizard` component manages a collection of child `WizardScreen` components, displays the correct wizard screen, displays errors if necessary, and handles required plugins.

Here is an example of a super simple 3-screen wizard (taken from the components demo):
```
<Wizard 
	activeScreen={ activeWizardScreen } 
	requiredPlugins={ [ 'jetpack' ] }
	requiredPluginsCancelText={ __( 'Back to checklist' ) }
	onRequiredPluginsCancel={ () => console.log( 'Checklist' ) }
>
	<WizardScreen
		identifier='test-wizard-1'
		completeButtonText={ __( 'Continue to 2' ) }
		onCompleteButtonClicked={ () => this.setState( { activeWizardScreen: 'test-wizard-2' } ) }
		subCompleteButtonText={ __( 'Back to checklist' ) }
		onSubCompleteButtonClicked={ () => console.log( 'Checklist' ) }
	>
		<FormattedHeader
			headerText={ __( 'WizardScreen Component 1' ) }
			subHeaderText={ __( 'This is #1 wizard screen' ) }
		/>
	</WizardScreen>
	<WizardScreen
		identifier='test-wizard-2'
		completeButtonText={ __( 'Continue to 3' ) }
		onCompleteButtonClicked={ () => this.setState( { activeWizardScreen: 'test-wizard-3' } ) }
		subCompleteButtonText={ __( 'Back to 1' ) }
		onSubCompleteButtonClicked={ () => this.setState( { activeWizardScreen: 'test-wizard-1' } ) }
	>
		<FormattedHeader
			headerText={ __( 'WizardScreen Component 2' ) }
			subHeaderText={ __( 'This is #2 wizard screen' ) }
		/>
	</WizardScreen>
	<WizardScreen
		noBackground
		identifier='test-wizard-3'
		completeButtonText={ __( 'Finish' ) }
		onCompleteButtonClicked={ () => this.setState( { activeWizardScreen: 'test-wizard-1' } ) }
	>
		<FormattedHeader
			headerText={ __( 'WizardScreen Component 3' ) }
			subHeaderText={ __( 'This is #3 wizard screen' ) }
		/>
	</WizardScreen>
</Wizard>
```
The `FormattedHeader` components are the screen contents; the children of the `WizardScreen` are the screen contents.

If there are changes required for this to be merged and people are blocked, feel free to push commits to this branch and handle them without me, since I'll be gone next week. I won't be offended. :)

This should hopefully be the last of the large, overarching PRs. After this is merged, we should be able to just have regular, single-focused PRs. 🤞 

### Next steps:
- We need to add handling for premium plugins to the `Plugin_Manager` class, so we have graceful failures when required premium plugins are missing. Currently, the plugin manager will try and download the plugins from .org and will probably crash (I haven't really tested what happens if you don't have e.g. WooCommerce Subscriptions installed at all with the required plugin handling)
- (Optional) We can easily set up a multi-page components demo now, with each component having its own `WizardScreen`. That would probably be preferable to our current giganto long list.

### How to test the changes in this Pull Request:

1. `npm run clean; npm run build:webpack`
2. Try the wizard in the Components Demo.
3. Try the Subscriptions and Subscriptions Onboarding wizards and verify things work well and look normal.

When reviewing the code, it's probably going to be super confusing to look at the diff for the Subscription and Subscription Onboarding wizards on GitHub. I'd advise looking at `wizards/subscriptions/index.js` and `wizards/subscriptionsOnboarding/index.js` to see what things look like using the new components.

### Other information:
